### ISSUE TITLE
[Enhancement]: how to extract photo to vector for towhee and torch to the same?

### ISSUE THREAD

**Issue by sulmz** [towhee-io/towhee#2404]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

I want to extract photo to vector on my GUI software
I cannot take a bag for towhee,but I can bag for torch
so, how can I use torch for extract photo to the same vector as for use towhee?

here is my current code:
use towhee:----------
import towhee
from towhee.functional.option import _Reason

from config import *

import numpy as np

from towhee.types import Image
import cv2
#图片解压
class ResNet50:
    def __init__(self):
        self.url_pipe = (towhee.dummy_input()
                    .image_decode()
                    .image_embedding.timm(model_name='resnet50')
                    .tensor_normalize()
                    .as_function()
        )
        self.byte_pipe = (towhee.dummy_input()
        .runas_op(self.decode)
        .image_embedding.timm(model_name='resnet50')
        .tensor_normalize()
        .as_function()
)

    def decode(self,content):
        arr = np.asarray(bytearray(content), dtype=np.uint8)
        return Image(cv2.imdecode(arr, -1), 'BGR')

    #图片url解压为向量
    def resnet50_extract_feat(self, img_path):
        feat = self.url_pipe(img_path)
        if isinstance(feat, _Reason):
            raise feat.exception
        return feat

    #图片本地路径解压为向量
    def extract_by_byte(self,img_path):
        with open(img_path,'rb')as f:
            content = f.read()
            vector = self.byte_pipe(content)
            return vector

# 主函数
if __name__ == '__main__':
    c = ResNet50()
    pic_path = r'/Users/sulimingzhou/Desktop/coding_study/search-image-client/assets/bannerRight12x.png'
    vector = c.extract_by_byte(img_path=pic_path)
    print('--vecotr',vector)

use torch-----------------
'''
图片解析为向量
'''
from torch.autograd import no_grad
from torch.serialization import load as torch_load
from torchvision import transforms
from torchvision import models
from PIL import Image
class Extract_Picture:
    '解压图片'
    def __init__(self,model_path=None) -> None:
        self.model_path = model_path
    # 加载图片，并进行预处理
    def preprocess_image(self,image_path):
        '加载图片，并进行预处理'
        # 读取图片并转换为张量
        image = Image.open(image_path).convert('RGB')
        # 定义预处理方式
        transform = transforms.Compose([
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        ])
        # 应用预处理
        image = transform(image).unsqueeze(0)
        return image

    # 加载模型，并解析图片
    def extract_feature(self,model, image):
        '加载模型，并解析图片'
        # 设置模型为评估模式
        model.eval()
        # 将图片输入模型，得到向量
        with no_grad():
            feature = model(image)
        return feature.squeeze().numpy()

    #解析本地图片为向量
    def extract_local_picture_to_vector(self,pic_path:str):
        '解析本地图片为向量'
        model = models.resnet50(weights=None)

        # 自定义模型存储位置
        model.load_state_dict(torch_load(self.model_path)) if self.model_path else None

        # 预处理图片
        img = self.preprocess_image(image_path=pic_path)

        # 解析图片
        vector = self.extract_feature(model=model,image=img)
        return vector

# 主函数
if __name__ == '__main__':
    # 图片路径
    image_path = r'D:\sulimingzhou\mingzhouyun\search-image-client\assets\128_黑_assets_167577938374445.jpg'
    # 自定义模型存储位置
    model_path = r'D:\sulimingzhou\mingzhouyun\search-image-client\resnet50_model\resnet50_a1_0-14fe96d1.pth'
    # 加载模型，并载入预训练权重
    c = Extract_Picture(model_path=model_path)
    feature = c.extract_local_picture_to_vector(pic_path=image_path)
    
    print('--vector',feature)

### Why is this needed?

extract photo to vector for towhee and torch to the same

### Anything else?

please change my torch code, please

**Comment by sulmz** [comment]
github change my code style，I rewrite it
`import towhee
from towhee.functional.option import _Reason

from config import *

import numpy as np

from towhee.types import Image
import cv2
#图片解压
class ResNet50:
    def __init__(self):
        self.url_pipe = (towhee.dummy_input()
                    .image_decode()
                    .image_embedding.timm(model_name='resnet50')
                    .tensor_normalize()
                    .as_function()
        )
        self.byte_pipe = (towhee.dummy_input()
        .runas_op(self.decode)
        .image_embedding.timm(model_name='resnet50')
        .tensor_normalize()
        .as_function()
)

    def decode(self,content):
        arr = np.asarray(bytearray(content), dtype=np.uint8)
        return Image(cv2.imdecode(arr, -1), 'BGR')

    #图片url解压为向量
    def resnet50_extract_feat(self, img_path):
        feat = self.url_pipe(img_path)
        if isinstance(feat, _Reason):
            raise feat.exception
        return feat

    #图片本地路径解压为向量
    def extract_by_byte(self,img_path):
        with open(img_path,'rb')as f:
            content = f.read()
            vector = self.byte_pipe(content)
            return vector

# 主函数
if __name__ == '__main__':
    c = ResNet50()
    pic_path = r'/Users/sulimingzhou/Desktop/coding_study/search-image-client/assets/bannerRight12x.png'
    vector = c.extract_by_byte(img_path=pic_path)
    print('--vecotr',vector)`




`'''
图片解析为向量
'''
from torch.autograd import no_grad
from torch.serialization import load as torch_load
from torchvision import transforms
from torchvision import models
from PIL import Image
class Extract_Picture:
    '解压图片'
    def __init__(self,model_path=None) -> None:
        self.model_path = model_path
    # 加载图片，并进行预处理
    def preprocess_image(self,image_path):
        '加载图片，并进行预处理'
        # 读取图片并转换为张量
        image = Image.open(image_path).convert('RGB')
        # 定义预处理方式
        transform = transforms.Compose([
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        ])
        # 应用预处理
        image = transform(image).unsqueeze(0)
        return image

    # 加载模型，并解析图片
    def extract_feature(self,model, image):
        '加载模型，并解析图片'
        # 设置模型为评估模式
        model.eval()
        # 将图片输入模型，得到向量
        with no_grad():
            feature = model(image)
        return feature.squeeze().numpy()

    #解析本地图片为向量
    def extract_local_picture_to_vector(self,pic_path:str):
        '解析本地图片为向量'
        model = models.resnet50(weights=None)

        # 自定义模型存储位置
        model.load_state_dict(torch_load(self.model_path)) if self.model_path else None

        # 预处理图片
        img = self.preprocess_image(image_path=pic_path)

        # 解析图片
        vector = self.extract_feature(model=model,image=img)
        return vector

# 主函数
if __name__ == '__main__':
    # 图片路径
    image_path = r'D:\sulimingzhou\mingzhouyun\search-image-client\assets\128_黑_assets_167577938374445.jpg'
    # 自定义模型存储位置
    model_path = r'D:\sulimingzhou\mingzhouyun\search-image-client\resnet50_model\resnet50_a1_0-14fe96d1.pth'
    # 加载模型，并载入预训练权重
    c = Extract_Picture(model_path=model_path)
    feature = c.extract_local_picture_to_vector(pic_path=image_path)
    
    print('--vector',feature)`

**Comment by sulmz** [comment]
github change my code style, I upload file
[search-image-client.zip](https://github.com/towhee-io/towhee/files/10719997/search-image-client.zip)

**Comment by junjiejiangjjj** [comment]
Because you normalize the vector in towhee.

![image](https://user-images.githubusercontent.com/14136703/218425653-29f63b90-3721-4222-8df0-d601e562154e.png)

**Comment by sulmz** [comment]
> Because you normalize the vector in towhee.
> 
> ![image](https://user-images.githubusercontent.com/14136703/218425653-29f63b90-3721-4222-8df0-d601e562154e.png)

Can I change my "torch" code rather than change my "towhee" code, to make sure the result vector is same?

**Comment by junjiejiangjjj** [comment]
`vec / np.linalg.norm(vec)`

**Comment by sulmz** [comment]
> `vec / np.linalg.norm(vec)`

thanks for reply，but still not the same, because the torch vector is  shape 1000.

**Comment by sulmz** [comment]
thanks for guorentong and M.Jael and 。,。，I got the final code
